### ISSUE TITLE
Refactor CvCity::doYields()

### ISSUE THREAD

**Issue by ShadesOT** [Religion-And-Revolution-Extended/Mod#75]

**Comment by ErikSandbraaten** [comment]
Please do not carry out this refactoring without  us agreeing on how to solve #87

**Comment by abmpicoli** [comment]
There is no description whatsoever of what this ticket is about... Refactor what, and to which purpose?

**Comment by abmpicoli** [comment]
Notice that these bugs are public to newcomers and the community: we should explain whenever possible what we intend to do...

**Comment by Nightinggale** [comment]
Changed milestone to 2.8 for multiple reasons. One is a suspicion that it breaks savegames and as such can't go in 2.7.x.

However we also need time to review the changes and test them. The amount of changed code grows and a proper review will take time. #196 doesn't help in finishing this quickly.

**Comment by ShadesOT** [comment]
Yeah, it is probably a good idea to not include this into a release until it is finished and tested. If time isn't our friend, ...

**Comment by ShadesOT** [comment]
The main refactoring comes along nicely.

But I spotted another possible bug in doYields() and it is one of those bugs, that stop me from progressing because I do not know what the game rules are.

The behavior in question is the rewarding of father trade points for auto-selling storage loss (part of the overflow) from a city to europe.
Auto-selling rewards father trade points only if the building, that is auto-selling, is the customs house or the central customs authority. Auto-selling because auf warehouse expansien 1&2 does not reward father trade points. Despite the fact, that the father points are calculated based on the profit. The profit already incorporates 
1. the percentage at which the storage loss is sold (and which is dependend on the building: warehouse expansion = 50%, warehouse expansion 2 = 60%, customs house = 80%, central customs authority = 90%)
2. the tax rate in europe

Is this the intended behavior?

I can not imagine it is. Getting father trade points for auto-selling is not documented to the players anywhere (is it?). But if a player assumes that auto-selling does reward father trade points (after all, it is selling to europe), then it should not depend on any particular building, but only if selling happened or not.
And code-wise it depends on, if the customs house trade settings have been unlocked by a building or not - which is not very generic.

I see 3 possibilities:
1. Keep it as it is.
Pro: does not change current game mechanic.
2. Do not reward father points for autoselling at all.
Pro: Players have to make more of an effort to get father trade points.
3. Reward father trade points for all auto-selling.
Pro: Coherent game mechanics.

How should we move along?

**Comment by abmpicoli** [comment]
In my opinion, Option 3. Reward father trade points for all auto-selling. Always.

**Comment by LibSpit** [comment]
I quite like 2.

**Comment by Nightinggale** [comment]
I would go for 3. Looks like everybody says either 2 or 3, which makes me say the solution would be to go for 3, but add some kind of on/off switch. You can hardcode it to on right now, but make it trivial to change later if we decide otherwise. That way you aren't stalled even if we don't disagree on a solution anytime soon.

**Comment by ShadesOT** [comment]
Boah, you are asking me to keep the quirks code? :-D

My vote goes to 3 as well.

We can think ahead meanwhile. Rewarding father trade points for all auto-selling raises some rebalancing questions:
1. Do we need to rebalance at all? The change would be the same for all human players.
2. The AI is only affected in the way that it loses a head start. Because the AI, as of now cheats hardcoded, in the way, that it has the customs house settings unlocked for all it's cities from the start. It does not understand how to use them, but as this example shows, they also have other benefits. From 2nd or 3rd lowest difficulty level on, the AI also gets a free base loss sell percentage ( = starting with a whopping 50% for not having any storage building at all). So maybe it starts gaining father trade points already as early as when there are 350 units of yield in the base camp. But I don't know if it ever is in the minor "uncomfortable" situation of having overflow and loss.

Maybe erik has some AI related thoughts on this topic.

**Comment by ErikSandbraaten** [comment]
I've always disliked the customs house \ sell-overflow since I think that it removes an essential part of the game which is to physically bring yields to Europe. However, I realize that we lack the automation that would be necessary to alleviate the increased burden of micromanagement if we were to choose a different system. The AI also likely benefits from not having to transport goods to Europe since it is generally inept when it comes to transport management and planning. Having said that, I clearly do not want to reward anyone for bypassing\short-circuiting this  important game system and thus I prefer option 2)

**Comment by Nightinggale** [comment]
`
Boah, you are asking me to keep the quirks code? :-D
`
No. If you are comparing solution 2 and 3, the difference is granting points. Now if you do "if (A==3) { grant points}", then we can use A to switch between 2 and 3 and as such we don't have to decide on hardcoding either of those solutions right now.

**Comment by ShadesOT** [comment]
I forgot to mention: Summarizing messages have a sound - which can be switched off. Messages never have a sound.
